pyquil/api/_quantum_computer.py

```diff
@@ -144,6 +144,13 @@ def run(self, executable: Executable,
             are a list of floats or integers.
         :return: A numpy array of shape (trials, len(ro-register)) that contains 0s and 1s.
         """
+        # This prevents a common error where users expect QVM.run()
+        # and QPU.run() to be interchangeable. QPU.run() needs the
+        # supplied executable to have been compiled, QVM.run() does not.
+        if isinstance(self.qam, QPU) and isinstance(executable, Program):
```

Comment:
why not put this check in the `QPU` class?

Reply:
From above:

>     1. The test is currently performed in QuantumComputer.run(). Should the test be performed in QPU.run()? I chose QC.run() because that is where the `executable` is provided and will show up in the backtrace.

Reply:
I think if you put it in QPU it will still show up in the backtrace, just one level up